File: sdk-py/iglobals_auth/jwks.py
import jwt
from typing import Dict, Any
from jwt import PyJWKClient

from .errors import ICAError

class JWKSService:
    def __init__(self, jwks_uri: str):
        # PyJWKClient provides caching out of the box
        self.jwks_client = PyJWKClient(jwks_uri, cache_keys=True)

    def verify_token(self, token: str, expected_aud: str, expected_iss: str) -> Dict[str, Any]:
        try:
            signing_key = self.jwks_client.get_signing_key_from_jwt(token)
            payload = jwt.decode(
                token,
                signing_key.key,
                algorithms=["RS256"],
                audience=expected_aud,
                issuer=expected_iss,
            )
            return payload
        except jwt.PyJWTError as e:
            raise ICAError('invalid_token', str(e), 401)

File: sdk-js/src/jwks.ts
import jwt, { JwtHeader, SigningKeyCallback } from 'jsonwebtoken';
import jwksClient from 'jwks-rsa';
import { JWTPayload } from './types';
import { ICAError } from './errors';

export class JWKSService {
  private client: jwksClient.JwksClient;

  constructor(jwksUri: string) {
    this.client = jwksClient({
      jwksUri,
      cache: true,
      cacheMaxEntries: 5,
      cacheMaxAge: 3600000, // 1 hour
    });
  }

  private getKey = (header: JwtHeader, callback: SigningKeyCallback) => {
    this.client.getSigningKey(header.kid, (err, key) => {
      if (err) {
        return callback(err);
      }
      const signingKey = key?.getPublicKey();
      callback(null, signingKey);
    });
  };

  public verifyToken(token: string, expectedAud: string, expectedIss: string): Promise<JWTPayload> {
    return new Promise((resolve, reject) => {
      jwt.verify(
        token,
        this.getKey,
        {
          audience: expectedAud,
          issuer: expectedIss,
          algorithms: ['RS256'],
        },
        (err, decoded) => {
          if (err) {
            reject(new ICAError('invalid_token', err.message, 401));
          } else {
            resolve(decoded as JWTPayload);
          }
        }
      );
    });
  }
}
